### PR TITLE
chore: improve `x/profiles` by mutation test

### PR DESCRIPTION
**Pull request by dadamu**
## Description

~~Deps: #1231~~
This PR improves `x/profiles` by mutation tests to increase mutation score of `x/profiles` to 0.887324.

```
################################################################################
### WARNING! This test will take hours to complete!
################################################################################
running mutation tests for the following module(s): profiles
writing mutation test result to mutation_test_result.txt
The mutation score is 0.887324 (252 passed, 2 failed, 0 duplicated, 30 skipped, total is 284)
```

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [ ] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: scripts/mutation-test.sh [hunk 1]
```diff
@@ -10,7 +10,7 @@ DISABLED_MUTATORS='branch/*'
 # * ignore test and Protobuf files
 # * ignore simulation test files
 # * ignore legacy files
-go_file_exclusions="-type f ! -path */client/* ! -path */simulation/* ! -path */legacy/* -name *.go -and -not -name *_test.go -and -not -name *pb* -and -not -name module.go -and -not -name ibc_module.go"
+go_file_exclusions="-type f ! -path */client/* ! -path */simulation/* ! -path */legacy/* ! -path */wasm/* -name *.go -and -not -name *_test.go -and -not -name *pb* -and -not -name module.go -and -not -name ibc_module.go"
```
Comment (dadamu) [inline scripts/mutation-test.sh:13]: */wasm/* files are legacy for back compatible for old `desmos-bindings`. Now the wasm contract related are all based on stargate feature.

File: scripts/mutation-test.sh [hunk 2]
```diff
@@ -10,7 +10,7 @@ DISABLED_MUTATORS='branch/*'
 # * ignore test and Protobuf files
 # * ignore simulation test files
 # * ignore legacy files
-go_file_exclusions="-type f ! -path */client/* ! -path */simulation/* ! -path */legacy/* -name *.go -and -not -name *_test.go -and -not -name *pb* -and -not -name module.go -and -not -name ibc_module.go"
+go_file_exclusions="-type f ! -path */client/* ! -path */simulation/* ! -path */legacy/* ! -path */wasm/* ! -path */testutil/* -name *.go -and -not -name *_test.go -and -not -name *pb* -and -not -name module.go -and -not -name ibc_module.go"
```
Comment (dadamu) [inline scripts/mutation-test.sh:13]: */wasm/* files are legacy and only for back compatible for old desmos-bindings. Now the wasm contract related are all based on stargate feature.

File: scripts/mockgen.sh [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
 
-find ./x -path "*$search_path" | while read -r source_file; do
-    destination_file="${source_file/types\/expected_keepers.go/testutil\/expected_keepers_mocks.go}"
-    destination_dir=$(dirname "$destination_file")
-    package=$(basename "$destination_dir")
+# Define an array of source and destination paths
+declare -a paths=(
+    "types/expected_keepers.go:testutil/expected_keepers_mocks.go"
+    "types/hooks.go:testutil/hooks_mocks.go"
+)
```
Comment (dadamu) [inline scripts/mockgen.sh:9]: Creating `hooks_mocks.go` can makes test keeper hooks easily

File: scripts/mutation-test.sh [hunk 4]
```diff
@@ -2,16 +2,18 @@
...
 # * ignore test and Protobuf files
 # * ignore simulation test files
 # * ignore legacy files
-go_file_exclusions="-type f ! -path */client/* ! -path */simulation/* ! -path */legacy/* -name *.go -and -not -name *_test.go -and -not -name *pb* -and -not -name module.go -and -not -name ibc_module.go"
-MUTATION_SOURCES=$(find ./x  $go_file_exclusions )
+# * ignore legacy wasm files
+# * ignore testutil files
+# * ignore codec.go
+go_file_exclusions="-type f -path */keeper/* -or -path */types/* -name *.go -and -not -path */legacy/* -and -not -name *_test.go -and -not -name *pb* -and -not -name codec.go"
+MUTATION_SOURCES=$(find ./x $go_file_exclusions )
```
Comment (dadamu) [inline scripts/mutation-test.sh:16]: Mutation tests should only focus on keeper, types and desmos-related features.

File: x/profiles/keeper/msg_server_dtag_transfers.go [hunk 5]
```diff
@@ -105,7 +105,7 @@ func (k MsgServer) AcceptDTagTransferRequest(goCtx context.Context, msg *types.M
 	}
 
 	// Get the current owner profile
-	currentOwnerProfile, exist, err := k.GetProfile(ctx, msg.Receiver)
```
Comment (dadamu) [inline x/profiles/keeper/msg_server_dtag_transfers.go:108]: renaming `currentOwnerProfile` into `requestReceiverProfile` makes it more explicit that focus on receiver of the transfer request.

File: x/profiles/keeper/msg_server_dtag_transfers.go [hunk 6]
```diff
@@ -116,31 +116,31 @@ func (k MsgServer) AcceptDTagTransferRequest(goCtx context.Context, msg *types.M
...
-	err = k.ValidateProfile(ctx, currentOwnerProfile)
+	requestReceiverProfile.DTag = msg.NewDTag
+	err = k.ValidateProfile(ctx, requestReceiverProfile)
 	if err != nil {
 		return nil, errors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
 	}
 
 	// Check for an existent profile of the receiving user
-	receiverProfile, exist, err := k.GetProfile(ctx, msg.Sender)
+	requestSenderProfile, exist, err := k.GetProfile(ctx, msg.Sender)
```
Comment (dadamu) [inline x/profiles/keeper/msg_server_dtag_transfers.go:132]: Renaming `receiverProfile` into `requestSenderProfile` makes it more explicit that focus on sender of the transfer request, it also avoid from confusing between the DTag receiver of the message and the receiver of the transfer request (they are different).

File: x/profiles/keeper/msg_server_app_link.go [hunk 7]
```diff
@@ -97,12 +97,6 @@ func (k MsgServer) UnlinkApplication(
 			sdk.NewAttribute(sdk.AttributeKeyAction, sdk.MsgTypeURL(msg)),
 			sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer),
 		),
-		sdk.NewEvent(
-			types.EventTypeApplicationLinkDeleted,
-			sdk.NewAttribute(types.AttributeKeyUser, msg.Signer),
-			sdk.NewAttribute(types.AttributeKeyApplicationName, msg.Application),
-			sdk.NewAttribute(types.AttributeKeyApplicationUsername, msg.Username),
-		),
```
Comment (dadamu) [inline x/profiles/keeper/msg_server_app_link.go:105]: It is duplicated with the event inside `DeleteApplicationLink`:
https://github.com/desmos-labs/desmos/blob/405f77bbe774ff3d2dd264f0f929959ffd1151cf/x/profiles/keeper/keeper_app_links.go#L109

File: x/profiles/keeper/msg_server_dtag_transfers.go [hunk 8]
```diff
@@ -157,29 +157,26 @@ func (k MsgServer) AcceptDTagTransferRequest(goCtx context.Context, msg *types.M
...
-	// Save the receiver profile
-	err = k.Keeper.SaveProfile(ctx, receiverProfile)
+	// Save the request sender profile
+	err = k.Keeper.SaveProfile(ctx, requestSenderProfile)
 	if err != nil {
 		return nil, err
 	}
 
-	k.DeleteAllUserIncomingDTagTransferRequests(ctx, msg.Receiver)
-	k.DeleteAllUserIncomingDTagTransferRequests(ctx, msg.Sender)
```
Comment (dadamu) [inline x/profiles/keeper/msg_server_dtag_transfers.go:181]: These are duplicate with the `storeProfileWithoutDTagCheck` called by `SaveProfile`:
https://github.com/desmos-labs/desmos/blob/405f77bbe774ff3d2dd264f0f929959ffd1151cf/x/profiles/keeper/keeper.go#L120-L121

File: x/profiles/types/account.go [hunk 9]
```diff
@@ -314,9 +314,14 @@ func (p *Profile) String() string {
 
 // MarshalYAML returns the YAML representation of a Profile.
 func (p *Profile) MarshalYAML() (interface{}, error) {
+	var pubKey = ""
+	if p.GetPubKey() != nil {
+		pubKey = p.GetPubKey().String()
+	}
```
Comment (dadamu) [inline x/profiles/types/account.go:320]: It avoid from the panicking since pubkey is `nil`

File: x/profiles/types/account_test.go [hunk 10]
```diff
@@ -39,21 +146,21 @@ func TestProfile_Update(t *testing.T) {
 			)),
 			update: types.NewProfileUpdate(
 				types.DoNotModify,
-				"",
+				types.DoNotModify,
```
Comment (dadamu) [inline x/profiles/types/account_test.go:149]: All the parameter in this test should be `DoNotModify`

File: x/profiles/types/models_params.go [hunk 11]
```diff
@@ -103,7 +103,7 @@ func ValidateNicknameParams(i interface{}) error {
 
 	// TODO make sense to cap this? I've done this thinking "what's the sense of having names higher that 1000 chars?"
 	maxLength := params.MaxLength
-	if maxLength.IsNil() || maxLength.IsNegative() || maxLength.GT(DefaultMaxNicknameLength) {
+	if maxLength.IsNil() || maxLength.LT(params.MinLength) {
```
Comment (dadamu) [inline x/profiles/types/models_params.go:106]: Improve `maxLentgh` to be not little than `minLength`

File: x/profiles/keeper/grpc_query.go [hunk 12]
```diff
@@ -96,12 +96,10 @@ func (k Keeper) ChainLinks(ctx context.Context, request *types.QueryChainLinksRe
...
 	linksPrefix := types.ChainLinksPrefix
 	switch {
-	case request.User != "" && request.ChainName != "" && request.Target != "":
-		linksPrefix = types.ChainLinksStoreKey(request.User, request.ChainName, request.Target)
 	case request.User != "" && request.ChainName != "":
-		linksPrefix = types.UserChainLinksChainPrefix(request.User, request.ChainName)
+		linksPrefix = types.ChainLinksStoreKey(request.User, request.ChainName, request.Target)
 	case request.User != "":
-		linksPrefix = types.UserChainLinksPrefix(request.User)
+		linksPrefix = types.UserChainLinksChainPrefix(request.User, request.ChainName)
```
Comment (dadamu) [inline x/profiles/keeper/grpc_query.go:102]: Improve the prefix that `types.UserChainLinksChainPrefix("user", "")` is equal to `types.UserChainLinkPrefix("user")`

File: x/profiles/keeper/grpc_query.go [hunk 13]
```diff
@@ -129,11 +127,8 @@ func (k Keeper) ChainLinkOwners(ctx context.Context, request *types.QueryChainLi
 	store := sdkCtx.KVStore(k.storeKey)
 
 	ownersPrefix := types.ChainLinkChainPrefix
-	switch {
-	case request.ChainName != "" && request.Target != "":
+	if request.ChainName != "" {
 		ownersPrefix = types.ChainLinkChainAddressKey(request.ChainName, request.Target)
```
Comment (dadamu) [inline x/profiles/keeper/grpc_query.go:131]: Similar as above, can be less one case since `types.ChainLinkChainAddressKey("chain_name", "")` is equal to `types.ChainLinkChainKey("chain_name")`

File: x/profiles/keeper/grpc_query.go [hunk 14]
```diff
@@ -166,11 +161,8 @@ func (k Keeper) DefaultExternalAddresses(ctx context.Context, request *types.Que
 	store := sdkCtx.KVStore(k.storeKey)
 
 	defaultPrefix := types.DefaultExternalAddressPrefix
-	switch {
-	case request.Owner != "" && request.ChainName != "":
+	if request.Owner != "" {
 		defaultPrefix = types.DefaultExternalAddressKey(request.Owner, request.ChainName)
```
Comment (dadamu) [inline x/profiles/keeper/grpc_query.go:165]: Similar as above

File: x/profiles/keeper/grpc_query.go [hunk 15]
```diff
@@ -203,12 +195,10 @@ func (k Keeper) ApplicationLinks(ctx context.Context, request *types.QueryApplic
 	// Get user links prefix store
 	linksPrefix := types.ApplicationLinkPrefix
 	switch {
-	case request.User != "" && request.Application != "" && request.Username != "":
```
Comment (dadamu) [inline x/profiles/keeper/grpc_query.go:206]: Similar as above

File: x/profiles/keeper/grpc_query.go [hunk 16]
```diff
@@ -253,11 +243,8 @@ func (k Keeper) ApplicationLinkOwners(ctx context.Context, request *types.QueryA
 	store := sdkCtx.KVStore(k.storeKey)
 
 	ownersPrefix := types.ApplicationLinkAppPrefix
-	switch {
-	case request.Application != "" && request.Username != "":
+	if request.Application != "" {
```
Comment (dadamu) [inline x/profiles/keeper/grpc_query.go:246]: Similar as above

File: x/profiles/keeper/grpc_query.go [hunk 17]
```diff
@@ -203,12 +195,10 @@ func (k Keeper) ApplicationLinks(ctx context.Context, request *types.QueryApplic
...
 	linksPrefix := types.ApplicationLinkPrefix
 	switch {
-	case request.User != "" && request.Application != "" && request.Username != "":
-		linksPrefix = types.UserApplicationLinkKey(request.User, request.Application, request.Username)
 	case request.User != "" && request.Application != "":
-		linksPrefix = types.UserApplicationLinksApplicationPrefix(request.User, request.Application)
+		linksPrefix = types.UserApplicationLinkKey(request.User, request.Application, request.Username)
 	case request.User != "":
-		linksPrefix = types.UserApplicationLinksPrefix(request.User)
+		linksPrefix = types.UserApplicationLinksApplicationPrefix(request.User, request.Application)
```
Comment (dadamu) [inline x/profiles/keeper/grpc_query.go:201]: Similar as above

File: x/profiles/keeper/keeper_chain_links_test.go [hunk 18]
```diff
@@ -326,8 +369,15 @@ func (suite *KeeperTestSuite) TestKeeper_GetChainLink() {
...
-	account := profilestesting.GetChainLinkAccount("cosmos", "cosmos")
-	secondAccount := profilestesting.GetChainLinkAccount("cosmos", "cosmos")
+	accounts := []profilestesting.ChainLinkAccount{
+		profilestesting.GetChainLinkAccount("cosmos", "cosmos"),
+		profilestesting.GetChainLinkAccount("cosmos", "cosmos"),
+		profilestesting.GetChainLinkAccount("cosmos", "cosmos"),
+	}
+	sort.SliceStable(accounts, func(i, j int) bool {
+		return accounts[i].Bech32Address().GetValue() < accounts[j].Bech32Address().GetValue()
+	})
```
Comment (dadamu) [inline x/profiles/keeper/keeper_chain_links_test.go:379]: It requires at least three chain links to test update oldest default external address:
https://github.com/desmos-labs/desmos/blob/405f77bbe774ff3d2dd264f0f929959ffd1151cf/x/profiles/keeper/keeper_chain_links.go#L88-L91

File: x/profiles/keeper/keeper.go [hunk 19]
```diff
@@ -242,7 +242,7 @@ func (k Keeper) ValidateProfile(ctx sdk.Context, profile *types.Profile) error {
 	}
 
 	maxBioLen := params.Bio.MaxLength.Int64()
-	if profile.Bio != "" && int64(len(profile.Bio)) > maxBioLen {
+	if int64(len(profile.Bio)) > maxBioLen {
```
Comment (dadamu) [inline x/profiles/keeper/keeper.go:245]: `profile.Bio != ""` is redundant since it means 0 length.

File: x/profiles/keeper/relay_app_links.go [hunk 20]
```diff
@@ -20,8 +20,8 @@ import (
 	oracletypes "github.com/desmos-labs/desmos/v6/x/oracle/types"
 )
 
-// oracleScriptCallData represents the data that should be OBI-encoded and sent to perform an oracle request
-type oracleScriptCallData struct {
+// OracleScriptCallData represents the data that should be OBI-encoded and sent to perform an oracle request
+type OracleScriptCallData struct {
```
Comment (dadamu) [inline x/profiles/keeper/relay_app_links.go:24]: make it public for testing

File: x/profiles/keeper/relay_chain_links.go [hunk 21]
```diff
@@ -61,13 +61,6 @@ func (k Keeper) OnRecvLinkChainAccountPacket(
 			hex.EncodeToString(profile.GetPubKey().Bytes()), hex.EncodeToString(pubKey.Bytes()))
 	}
 
-	// Verify the source proof by making sure the user has signed the destination
-	// address using the source address private key
-	err = data.SourceProof.Verify(k.cdc, k.legacyAmino, data.DestinationAddress, srcAddrData)
-	if err != nil {
-		return packetAck, err
-	}
```
Comment (dadamu) [inline x/profiles/keeper/relay_chain_links.go:69]: This check is duplicated with `SaveChainLink` to verify the source proof:
https://github.com/desmos-labs/desmos/blob/20e9380bd9ac0847ac1a0b3de71fd1cde598e29d/x/profiles/keeper/keeper_chain_links.go#L38-L39

File: x/profiles/types/models_params.go [hunk 22]
```diff
@@ -140,11 +140,11 @@ func ValidateDTagParams(i interface{}) error {
 		return fmt.Errorf("empty dTag regEx param")
 	}
 
-	if params.MinLength.IsNegative() || params.MinLength.LT(DefaultMinDTagLength) {
+	if params.MinLength.LT(DefaultMinDTagLength) {
```
Comment (dadamu) [inline x/profiles/types/models_params.go:143]: `DefaultMinDTagLength` is 2, so it must not be negative number

File: x/profiles/keeper/msg_server_dtag_transfers.go [hunk 23]
```diff
@@ -154,32 +153,29 @@ func (k MsgServer) AcceptDTagTransferRequest(goCtx context.Context, msg *types.M
 
 		senderAcc := k.ak.GetAccount(ctx, add)
 		if senderAcc == nil {
-			senderAcc = authtypes.NewBaseAccountWithAddress(add)
+			senderAcc = k.ak.NewAccountWithAddress(ctx, add)
```
Comment (dadamu) [inline x/profiles/keeper/msg_server_dtag_transfers.go:156]: It would be a problem using `NewBaseAccountWithAddress` since its account number is always 0.
To solve it, `accountKeeper.NewAccountWithAddress` is the proper way to create an account

File: x/profiles/types/models_chain_links_test.go [hunk 24]
```diff
@@ -920,7 +1008,7 @@ func TestHexAddress_Validate(t *testing.T) {
 		},
 		{
 			name:      "address value shorter than prefix returns error",
-			address:   types.NewHexAddress("0", "0x"),
+			address:   types.NewHexAddress("0x", "0x"),
```
Comment (dadamu) [inline x/profiles/types/models_chain_links_test.go:1011]: Using same length tests the boundary condition of this:
https://github.com/desmos-labs/desmos/blob/b20b8c895232a9372a45c58395df0cb808998a39/x/profiles/types/models_chain_links.go#L603C5-L603C34
